### PR TITLE
kvclient: ignore stale lease information from lagging replicas

### LINKED ISSUE #72772: kvclient: range descriptor cache can thrash during range rebalancing
When the `DistSender` routes a `BatchRequest` to a replica and receives a `NotLeaseHolderError`, it checks if the `Lease` or the `LeaseHolder` fields of the error (if populated) point to a replica that is compatible with the cached range descriptor. If not, it assumes that the cached range descriptor must be stale and triggers a cache eviction and performs a range descriptor lookup from KV. 
https://github.com/cockroachdb/cockroach/blob/50a491d4bae701fbba31293810864fb79189250f/pkg/kv/kvclient/rangecache/range_cache.go#L1320-L1326

This turns out to be a flawed assumption because its also possible that the cached range descriptor is not stale but that the replica returning the NLHE is lagging behind the rest of the range and is operating under an older leaseholder that has since been removed from the range. 

As a result of this bug, until the replica being routed to learns about a newer leaseholder that _is_ part of the range, it will keep returning an NLHE and will cause the `DistSender` to keep performing range lookups & to keep refreshing the cached range descriptor. This means that in the presence of background range rebalancing in the cluster, we risk `BatchRequest`s being ping-ponged back and forth from the `DistSender` to a lagging replica and thrashing at the range cache. We saw this play out in a customer escalation where tail latencies spiked up to dozens of seconds during a rolling decommissioning cycle of their cluster. 

cc @cockroachdb/kv

### PR DESCRIPTION
**Pull request by aayushshah15**
This commit makes it such that the DistSender ignores the lease information
returned in `NotLeaseHolderError`s coming from replicas that have stale view of
the range descriptor (characterized by an older `DescriptorGeneration` on the
replica).

Not doing so before was hazardous because, if we received an NLHE that pointed
to a replica that did not belong in the cached descriptor, we'd trigger a cache
evicion. This assumed that the replica returning the error had a fresher view
of the range than what we had in the cache, which is not always true. This
meant that we'd keep doing range lookups and subsequent evictions until this
lagging replica caught up to the current state of the range.

Fixes #72772

Release note (bug fix): A bug that caused high SQL tail latencies during
background rebalancing in the cluster has been fixed.

### CODE REVIEW

**Comment by aayushshah15** [comment]
Note: In this commit, we're _always_ ignoring the lease information coming from replicas that have a stale view of the range descriptor, even if the lease information is compatible with the cached range descriptor. I wasn't sure if using the lease information from a lagging replica is better than simply trying the next replica in the cached range descriptor. Do y'all have thoughts on that?

**Review by nvb — COMMENTED** [review]
> Note: In this commit, we're always ignoring the lease information coming from replicas that have a stale view of the range descriptor, even if the lease information is compatible with the cached range descriptor. I wasn't sure if using the lease information from a lagging replica is better than simply trying the next replica in the cached range descriptor. Do y'all have thoughts on that?

This is an interesting observation. It relates to a comment I was going to leave down below, but which I'm going to pull up here instead.

We're adding conditional logic to compare the information in the `NotLeaseHolderError` with the information in the eviction token up in `sendToReplicas`. This isn't quite how things work elsewhere. In most places, this kind of logic lives in the `EvictionToken` or `CacheEntry`. The methods on `EvictionToken` like `UpdateLease` and `EvictAndReplace` deal with the messy job of trying to reconcile partial and inconsistent information into mostly-monotonic state transitions in the cache.

If we follow the model laid out there, then I think the approach we'd want to take with this change is to plumb the descriptor generation through `EvictionToken.UpdateLease` and `EvictionToken.UpdateLeaseholder` into `CacheEntry.updateLease`. Inside `CacheEntry.updateLease`, we could then update [the logic that consults the range descriptor](https://github.com/cockroachdb/cockroach/blob/1ce529271fd22077fd196ba41c76d5076be93ad5/pkg/kv/kvclient/rangecache/range_cache.go#L1327-L1333) to look something like:
```go
_, ok := e.desc.GetReplicaDescriptorByID(l.Replica.ReplicaID)
if !ok {
    // ...
    if gen != 0 && gen < e.desc.Generation {
        return false, e
    }
    // ...
    return true, nil
}
```

One major benefit of this is that it will be easier to test (in addition to the high-level test you already have).

> Reviewed 3 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-:-Mqv5gqfAIUobzmuResQ:b-qnwf5h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 2086 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv2PU85rSyd4u23yM7:-Mqv2PU85rSyd4u23yM8:b-bjcse) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvclient/kvcoord/dist_sender.go#L2086)):*
> ```Go
> 				// We cannot trust the lease information in the error if it is coming
> 				// from a replica that has a stale view of the range (compared to what
> 				// we've got in the cache).
> ```

Could you add to this comment and explain how we can get in this situation, what it means regarding the state of the replica that returned this error, and why we're careful not to evict from the cache in this case?
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 2094 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv378_An8XRsbm9dF2:-Mqv378_An8XRsbm9dF3:b-5imbi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvclient/kvcoord/dist_sender.go#L2094)):*
> ```Go
> 					)
> 				}
> 				if (tErr.Lease != nil || tErr.LeaseHolder != nil) && !withStaleRangeDesc {
> ```

nit: consider avoiding the nesting and instead splitting this into multiple conditions with break statements and explanations in each. Something like:
```go
case *roachpb.NotLeaseHolderError:
    ds.metrics.NotLeaseHolderErrCount.Inc(1)

    if tErr.Lease == nil && tErr.LeaseHolder == nil {
        // The replica knew it did not hold the lease, but did not provide us
        // with a better alternative. Loop around and try the next replica.
        break
    }

    if tErr.DescriptorGeneration != 0 && tErr.DescriptorGeneration < routing.Desc().Generation {
        // The replica had a stale view of the range ...
        break
    }

    // If we got some lease information, we use it to update our cache.
    // Naively this would also happen when the next RPC comes back, but we
    // don't want to wait out the additional RPC latency.
    var updatedLeaseholder bool
```
___
*[pkg/kv/kvclient/kvcoord/dist_sender_test.go, line 1023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv8nTHClhPn6gBNXFO:-Mqv8nTHClhPn6gBNXFP:bkghjdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L1023)):*
> ```Go
> 			return reply, nil
> 		}
> 		require.Equal(t, ba.Replica.NodeID, roachpb.NodeID(1))
> ```

nit: instead of asserting the node ID here and then maintaining a `callsToNode2` variable, consider creating a `var calls []int` variable, append the node ID on each call to `sendFn`, and then asserting that the slice looks like `[]int{2, 1}` at the end.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 908 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv1tzl84hmpQNdFJHA:-Mqv1tzl84hmpQNdFJHB:b51jxy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvserver/replica_range_lease.go#L908)):*
> ```Go
> 		RangeID:              rangeDesc.RangeID,
> 		CustomMsg:            msg,
> 		DescriptorGeneration: rangeDesc.Generation,
> ```

nit: add this above `CustomMsg` to keep the fields in the same order as their declaration.
___
*[pkg/roachpb/errors.proto, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv0nbdCb-sl86y7X1X:-Mqv0nbdCb-sl86y7X1Y:bylr25p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/roachpb/errors.proto#L60)):*
> ```protobuf
>       (gogoproto.customname) = "RangeID", (gogoproto.casttype) = "RangeID"];
>   // The range descriptor generation of the replica the error originated from.
>   // Used to determine whether the error was returned because the replica had a
> ```

"Used by the DistSender to determine"...

**Review by aayushshah15 — COMMENTED** [review]
I changed this around to do this reconciliation inside `CacheEntry.updateLease()` and added a lower level test check for it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-:-MtbgbXs6KSqTL25_suI:b-u9og5f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @andreimatei, and @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 2086 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv2PU85rSyd4u23yM7:-MtbgWHz3JmP0jGIEsg8:b-1kmhwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvclient/kvcoord/dist_sender.go#L2086)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add to this comment and explain how we can get in this situation, what it means regarding the state of the replica that returned this error, and why we're careful not to evict from the cache in this case?
</blockquote></details>

Added this explanation inside `updateLease`.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 908 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv1tzl84hmpQNdFJHA:-Mtbf0am9VomeES6DHzj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvserver/replica_range_lease.go#L908)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: add this above `CustomMsg` to keep the fields in the same order as their declaration.
</blockquote></details>

Done.
___
*[pkg/roachpb/errors.proto, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv0nbdCb-sl86y7X1X:-Mtbf-bQ7z-M3H4JlZ6U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/roachpb/errors.proto#L60)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"Used by the DistSender to determine"...
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_test.go, line 1023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-Mqv8nTHClhPn6gBNXFO:-MtbgTTrDAfVjoZ2zi6p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31a0692cab5349975f5c35aaa60a44707734ab1/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L1023)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: instead of asserting the node ID here and then maintaining a `callsToNode2` variable, consider creating a `var calls []int` variable, append the node ID on each call to `sendFn`, and then asserting that the slice looks like `[]int{2, 1}` at the end.
</blockquote></details>

Done.

**Review by nvb — COMMENTED** [review]
This all :lgtm:, but before merging it I think @arulajmani should take a look.

Arul has been investigating a similar issue where the `NotLeaseHolderError` comes from a replica with a newer view of the RangeDescriptor than that which is in the RangeCache. In other words, he's been exploring an issue that looks to be caused by the inverse `gen > e.desc.Generation` variant of this "the lease is incompatible with the cached descriptor" check. The solution to avoiding a client-side backoff in that case may be very similar to the one here. However, this is a case where the newer `RangeDescriptor` would be valuable to the range cache. So his issue may motivate a small change in this PR where we add the full `RangeDescriptor` to the `NotLeaseHolderError`, instead of just the `RangeGeneration`. We could then use this updated `RangeDescriptor` to directly update the `RangeCache`, avoiding a cache eviction or meta lookup.

> Reviewed 4 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-:-MuBn5wv-kdQ8wLlbo-K:b2quefg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @andreimatei, and @nvanbenschoten)

**Comment by aayushshah15** [comment]
> In other words, he's been exploring an issue that looks to be caused by the inverse gen > e.desc.Generation variant of this "the lease is incompatible with the cached descriptor" check

To check my understanding, in that instance, the range cache currently thinks that the lease _is_ compatible with the cached descriptor even when the lease is on a replica that's a `LEARNER` in the cached desc. 
https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1352-L1356

So, despite the fact that we're returning `true` from `EvictionToken.UpdateLease()`, since the `LEARNER` replica isn't part of the transport, we end up trying all the other replicas, and _then_ finally do a cache eviction. 
https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/kvcoord/dist_sender.go#L1933-L1938

Is this what's happening in that issue?

**Comment by arulajmani** [comment]
> So, despite the fact that we're returning true from EvictionToken.UpdateLease(), since the LEARNER replica isn't part of the transport, we end up trying all the other replicas, and then finally do a cache eviction.

In that issue we see backoffs for every replica that we try, and as backoffs are predicated on `EvictionToken.UpdateLeases()` returning false, I don't think we end up reaching the first TODO you've linked above. I think what's happening here is that the `RangeCache` has been updated with the correct lease information between the time we constructed the transport and got the NLHE error. When we sync the eviction token we update it with the new lease information. Then, when we try to update the lease with the one returned by the NLHE, we think they're the same and return false (in `updateLease`).

https://github.com/cockroachdb/cockroach/blob/01cb84707dd8a739c7b8aa0587383e8dac71507d/pkg/kv/kvclient/rangecache/range_cache.go#L387-L394

We end up trying all replicas because the leasheholder on the routing is not a part of the transport, so I think the second TODO you linked above does partially apply. It's this logic here:
https://github.com/cockroachdb/cockroach/blob/d0b0d4528292966818c8d78fe9c31c239bbf17a0/pkg/kv/kvclient/kvcoord/dist_sender.go#L2114-L2126

We can trace this through together if you want. If we detect this `gen > e.desc.Generation` issue, we can short circuit trying all the replicas in the transport, update the range descriptor in the range cache, and retry with a new transport constructed off the new lease information + descriptor. 

I'll review this in a bit.

**Review by arulajmani — COMMENTED** [review]
The change looks good to me as well. I was thinking a bit more about the `gen > e.desc.Generation` case a bit more and we may not need to ship the entire `RangeDescriptor` on the NLHE to make it work. As long as we can identify that we're in that scenario, avoid backing off on the client, and retry, we might be in the clear -- this is because how we construct the slice of replicas to create the transport with:

https://github.com/cockroachdb/cockroach/blob/acdda53fc2b09d159c432284c64728ef0a40d22f/pkg/kv/kvclient/kvcoord/replica_slice.go#L61-L67

The targeted approach there would be to identify this scenario and make sure the lease information in the `RangeCache` is up-to-date (which is already happening) before we retry. That being said, it might still be beneficial to ship the entire `RangeDescriptor` and opportunistically update the `RangeCache` when we can. I'll let you guys decide if we should do this as part of this patch or not. 

> Reviewed 1 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-:-MuColvo0ps8Qv-1_6dH:b-i3twyl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @andreimatei, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-MuCmgnxDN9aBmIHf_7U:-MuCmgnxDN9aBmIHf_7V:bcq2t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1310)):*
> ```Go
> // entry.
> //
> // If the new leaseholder is not a replica in the descriptor, we assume the
> ```

Should we re-word this too given the change?
___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1336 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-MuCmLdm7cZ7i5JKiZoE:-MuCmLdm7cZ7i5JKiZoF:bwydlym) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1336)):*
> ```Go
> 
> 	// If the lease is incompatible with the cached descriptor and the error is
> 	// coming from a replica that has a non-stale descriptor, the cached
> ```

nit: s/non-stale/newer/

**Review by aayushshah15 — COMMENTED** [review]
@arulajmani and I were on a call to stare at the code and understand how things were going wrong in that issue he was investigating. We have a better understanding of the issue now (at least I think Arul does), but there's enough going on here that we think it should be discussed separately in its own issue / patch. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-:-MuRpR0m5u3sIlQXSdnU:br6mvdn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @andreimatei, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-MuCmgnxDN9aBmIHf_7U:-MuRqVx4DDexaMu-xibI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1310)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we re-word this too given the change?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1336 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-MuCmLdm7cZ7i5JKiZoE:-MuRpvEU0BAAFvKgy7cP:banfxjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1336)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/non-stale/newer/
</blockquote></details>

I do mean `non-stale` here. We don't need the error to come from a replica that has a _newer_ descriptor, it just needs to not be older than the one in the cache.

**Review by arulajmani — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73697)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @andreimatei, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1336 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73697#-MuCmLdm7cZ7i5JKiZoE:-MuRxEpuChEU29rWRu94:b-nqvf1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf7572915dbfb6b0d0737a0083cecfab6570365/pkg/kv/kvclient/rangecache/range_cache.go#L1336)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I do mean `non-stale` here. We don't need the error to come from a replica that has a _newer_ descriptor, it just needs to not be older than the one in the cache.
</blockquote></details>

I think we're saying the same thing, but I was reading `non-stale` as "newest" in absolute terms.

**Comment by aayushshah15** [comment]
TFTR!!

bors r+

**Comment by aayushshah15** [comment]
Should we backport this to 21.1 and 21.2 or should we wait until we have a fix for the other issue we were discussing on this PR?

**Pull request merged**
